src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/AutoClosingTagOnAutoInsertProvider.cs

```diff
@@ -0,0 +1,201 @@
+        {
+            if (optionsMonitor is null)
+            {
+                throw new ArgumentNullException(nameof(optionsMonitor));
+            }
+
+            _optionsMonitor = optionsMonitor;
+        }
+
+        public override string TriggerCharacter => ">";
+
+        public override bool TryResolveInsertion(Position position, FormattingContext context, out TextEdit edit, out InsertTextFormat format)
+        {
+            if (position is null)
+            {
+                throw new ArgumentNullException(nameof(position));
+            }
+
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (!_optionsMonitor.CurrentValue.AutoClosingTags)
+            {
+                format = default;
+                edit = default;
+                return false;
+            }
+
+            if (!TryGetTagInformation(context, position, out var tagName, out var autoClosingBehavior))
+            {
+                format = default;
+                edit = default;
+                return false;
+            }
+
+            format = InsertTextFormat.Snippet;
+            if (autoClosingBehavior == AutoClosingBehavior.EndTag)
+            {
+                edit = new TextEdit()
+                {
+                    NewText = $"$0</{tagName}>",
+                    Range = new Range(position, position)
+                };
+            }
+            else
+            {
+                Debug.Assert(autoClosingBehavior == AutoClosingBehavior.SelfClosing);
+
+                // Need to replace the `>` with ' />$0'
+                var replacementRange = new Range(
+                    start: new Position(position.Line, position.Character - 1),
+                    end: position);
+                edit = new TextEdit()
+                {
+                    NewText = " />$0",
+                    Range = replacementRange
+                };
+
+            }
+
+            return true;
+        }
+
+        private static bool TryGetTagInformation(FormattingContext context, Position position, out string name, out AutoClosingBehavior autoClosingBehavior)
+        {
+            var syntaxTree = context.CodeDocument.GetSyntaxTree();
+
+            var absoluteIndex = position.GetAbsoluteIndex(context.SourceText) - 1;
+            var change = new SourceChange(absoluteIndex, 0, string.Empty);
+            var owner = syntaxTree.Root.LocateOwner(change);
+            if (owner?.Parent == null)
+            {
+                name = null;
+                autoClosingBehavior = default;
+                return false;
+            }
+
+            if (owner.Parent is MarkupStartTagSyntax startTag &&
+                startTag.Parent is MarkupElementSyntax)
+            {
+                name = startTag.Name.Content;
+                autoClosingBehavior = InferAutoClosingBehavior(name);
+                name = startTag.GetTagNameWithOptionalBang();
```

Comment:
Kinda interesting that `name` is assigned to twice here. For clarity I'd consider making the first occurrence assigned to its own variable, unless there's an underlying reason for doing it this way?

Reply:
No underlying reason, I like your suggestion 😄